homeassistant/components/ondilo_ico/api.py

```diff
@@ -45,8 +45,33 @@ def get_all_pools_data(self) -> list[dict[str, Any]]:
             _LOGGER.debug(
                 "Retrieving data for pool/spa: %s, id: %d", pool["name"], pool["id"]
             )
-            pool["ICO"] = self.get_ICO_details(pool["id"])
-            pool["sensors"] = self.get_last_pool_measures(pool["id"])
-            _LOGGER.debug("Retrieved the following sensors data: %s", pool["sensors"])
 
-        return pools
+            try:
+                pool["ICO"] = self.get_ICO_details(pool["id"])
+                if not pool["ICO"]:
+                    _LOGGER.error("The pool id %s does not have any ICO attached")
+                    continue
+            except OndiloError as exc:
+                _LOGGER.error(
+                    "Error retrieving ICO details of your pool id %d. Server error was: %s",
+                    pool["id"],
+                    exc,
+                )
+                continue
+
+            try:
+                pool["sensors"] = self.get_last_pool_measures(pool["id"])
+                _LOGGER.debug(
+                    "Retrieved the following sensors data: %s", pool["sensors"]
+                )
+            except OndiloError as exc:
+                # Trying to retrieve data from an ICO that was replaced produces an error
+                # So we'll remove the pool/spa from the list
+                _LOGGER.error(
+                    "Error retrieving last data of your ICO '%s'. Server error was: %s",
+                    pool["ICO"],
+                    exc,
+                )
+                pool["ICO"] = None
+
+        return [pool for pool in pools if pool.get("ICO")]
```

Comment:
Let's not remove items from the list like this, rather create a new list and add items dat have everything

Reply:
Not sure I understand what you mean here. This is exactly what is happening: it creates a new list from the existing one and returns it.